host_modules/image_service.py

```diff
@@ -0,0 +1,70 @@
+"""
+Services related to SONiC images, such as:
+1) Download
+2) Install
+"""
+
+import logging
+import subprocess
+import os
+
+from host_modules import host_service
+
+MOD_NAME = "image_service"
+
+DEFAULT_IMAGE_SAVE_AS = "/host/downloaded-sonic"
+
+logger = logging.getLogger(__name__)
+
+
+class ImageService(host_service.HostModule):
+    """DBus endpoint that handles downloading and installing SONiC images"""
+
+    @host_service.method(
+        host_service.bus_name(MOD_NAME), in_signature="ss", out_signature="is"
+    )
+    def download(self, image_url, save_as):
+        """
+        Download a SONiC image.
+
+        Args:
+             image_url: url for remote image.
+             save_as: local path for the downloaded image
+        """
+        logger.info("Download new sonic image from {} as {}".format(image_url, save_as))
+        # Create parent directory.
+        dir = os.path.dirname(save_as)
+        if not os.path.exists(dir):
+            os.makedirs(dir)
```

Comment:
>makedirs [](http://example.com/codeflow?start=15&length=8)

This function will be finally called by outside, and it is an attacking point. Can we enforce that the dir must exist, and every user has the "w" permission, similar to "/tmp" permission ?

Reply:
Done.